app/src/org/commcare/models/database/IndexedFixturePathUtils.java

```diff
@@ -19,34 +24,65 @@
-        Cursor c = db.query(IndexedFixturePathsConstants.INDEXED_FIXTURE_PATHS_TABLE,
-                new String[]{IndexedFixturePathsConstants.INDEXED_FIXTURE_PATHS_COL_BASE, IndexedFixturePathsConstants.INDEXED_FIXTURE_PATHS_COL_CHILD},
-                IndexedFixturePathsConstants.INDEXED_FIXTURE_PATHS_COL_NAME + "=?", new String[]{fixtureName}, null, null, null);
+    public final static String INDEXED_FIXTURE_PATHS_TABLE = "IndexedFixtureIndex";
+    public final static String INDEXED_FIXTURE_PATHS_COL_NAME = "name";
+    public final static String INDEXED_FIXTURE_PATHS_COL_BASE = "base";
+    public final static String INDEXED_FIXTURE_PATHS_COL_CHILD = "child";
+    public final static String INDEXED_FIXTURE_PATHS_COL_ATTRIBUTES = "attributes";
+
+    public final static String INDEXED_FIXTURE_PATHS_TABLE_STMT =
+            "CREATE TABLE IF NOT EXISTS " +
+                    INDEXED_FIXTURE_PATHS_TABLE +
+                    " (" + INDEXED_FIXTURE_PATHS_COL_NAME + " UNIQUE" +
+                    ", " + INDEXED_FIXTURE_PATHS_COL_BASE +
+                    ", " + INDEXED_FIXTURE_PATHS_COL_CHILD +
+                    ", " + INDEXED_FIXTURE_PATHS_COL_ATTRIBUTES + ");";
+
+    public final static String INDEXED_FIXTURE_INDEXING_STMT =
+            DatabaseIndexingUtils.indexOnTableCommand("fixture_name_index",
+                    INDEXED_FIXTURE_PATHS_TABLE, INDEXED_FIXTURE_PATHS_COL_NAME);
+
+    public final static String INDEXED_FIXTURE_PATHS_TABLE_STMT_V15 =
+            "CREATE TABLE IF NOT EXISTS " +
+                    INDEXED_FIXTURE_PATHS_TABLE +
+                    " (" + INDEXED_FIXTURE_PATHS_COL_NAME + " UNIQUE" +
+                    ", " + INDEXED_FIXTURE_PATHS_COL_BASE +
+                    ", " + INDEXED_FIXTURE_PATHS_COL_CHILD + ");";
+
+    public static IndexedFixtureIndex lookupIndexedFixturePaths(SQLiteDatabase db,
+                                                                String fixtureName) {
+        Cursor c = db.query(INDEXED_FIXTURE_PATHS_TABLE,
+                new String[]{INDEXED_FIXTURE_PATHS_COL_BASE, INDEXED_FIXTURE_PATHS_COL_CHILD, INDEXED_FIXTURE_PATHS_COL_ATTRIBUTES},
+                INDEXED_FIXTURE_PATHS_COL_NAME + "=?", new String[]{fixtureName}, null, null, null);
         try {
             if (c.getCount() == 0) {
                 return null;
             } else {
                 c.moveToFirst();
-                return Pair.create(
-                        c.getString(c.getColumnIndexOrThrow(IndexedFixturePathsConstants.INDEXED_FIXTURE_PATHS_COL_BASE)),
-                        c.getString(c.getColumnIndexOrThrow(IndexedFixturePathsConstants.INDEXED_FIXTURE_PATHS_COL_CHILD)));
+                TreeElement attrsHolder = null;
+                byte[] attrsBlob = c.getBlob(c.getColumnIndexOrThrow(INDEXED_FIXTURE_PATHS_COL_ATTRIBUTES));
+                if (attrsBlob != null) {
+                    attrsHolder = SerializationUtil.deserialize(attrsBlob, TreeElement.class);
```

Comment:
Additionally, I think we should have a concrete exception handling strategy for issues here, since deserialization exceptions are probably a very unexpected outcome.

It *might* be worth doing something like retrieving the byte[] here, and letting the TreeElement get deserialized at a later point (like in get() here: https://github.com/dimagi/commcare-core/pull/842/files#diff-de4be0557361343fe019d6c9b25d8598R38) when we are actively doing reads/deserializations from storage more broadly?

Reply:
> Additionally, I think we should have a concrete exception handling strategy for issues here, since deserialization exceptions are probably a very unexpected outcome.

Do you mean that we don't want the app to crash on such deserialization exceptions and instead show some sort of error message to the user ?

Reply:
Mostly I mean that having the app crash here in a deserialization exception would be extremely confusing because it would be a symptom of an event that's out of the user's control and would be handled in a very strange part of the lifecycle.

Generally speaking I think it's very, very dangerous for "Utility" classes like this which are static and effectively stateless to throw errors that aren't strongly meaningful at the time. To the programmer, this method looks like it's just receiving a db hook and a string and doing some lookups. The caller wouldn't be expecting to need to be ready to handle mismatched serialized internal models based on the signature and expectations.

That **would** be their expectation later, however, when they are actually managing the resulting lookup table, so pushing the deserialization code there might be cleaner.

Reply:
@ctsims Thanks. I get your point now. Though I am struggling to find a better place for the deserialization to go since it requires access to the Prototype factory and hence needs to be in the `commcare-android`. The `get` function you linked above is part of `commcare-core`. The only other place I can forsee is 1 level up i.e. [getIndexedFixtureIdentifier](https://github.com/dimagi/commcare-android/blob/last_sync/app/src/org/commcare/models/database/AndroidSandbox.java#L67) but it again has the same semantics issue that doesn't convey to the caller to handle deserialization. 

Alternatively, we can decide to not catch `DeserializationException` in `SerializationUtil.deserialize` which would add a `throws DeserializationException` into function signature and hence will be much more evident to a caller function.

Reply:
Hm, I see what you are saying. 

I can't think of a great workaround for now, and it is true that if we broke the TreeElement serialization, we'd need to be fixing the records from scratch anyway. 

Maybe the big point of confusion here is just that it doesn't feel right for the attribute holder to be in this record to begin with. Can we make a second version of this method that **doesn't** pull the attrsHolder so that if we needed to just get the base and child paths, that there's a path for that? Would be good for this version to also throw a typed error specific to this problem so the end user  (and us) can see very deliberately that it's an issue with the stored format for retrieving this.

I'm thinking about what pieces we'll need to have in place in a circumstance where we need  to migrate fixtures due to a TreeElement serialization format change, and right now this method would just be a blocker.

Reply:
@ctsims 

That makes a lot of sense though just to clarify are you suggesting that we create an unused function/ code path right now because we might need it in future ?